### PR TITLE
Mega-patch to pull in various updates from forks

### PR DESCRIPTION
**Pull request by fhunleth**
[Apologies for the mega-patch. I think that this will be worthwhile change, but it's probably worth tagging the project to avoid disrupting anyone using it by a pretty major refactoring. Also, the diff view on GitHub is a little scary due to the deletions. It looks more recognizable as Erlang/ALE when you look at my repository.]

Many of the updates come from changes to the C code to fix race
conditions, simplify building (including cross-compiling), and add
functionality. The changes had been collected in the Elixir/ALE project,
and given that both projects had similar issues, it seemed better to
hold the fixes here. Given the nature of the changes, it was difficult
to commit incrementally and still have working builds. Hence this
mega-patch.

This code has been manually tested on an Erlang Embedded Demo Board.

The following is a summary of changes:

  1. Updated documentation with schematics showing use
  2. Fixed race conditions with GPIO (no more multiple threads in C)
  3. Fix interrupt support on Raspberry Pi 2. No more infinite
     interrupts.
  4. Eliminate dependencies. This made is possible to build in some
     situations. It also should help people use it out of the box.
  5. Use binaries on I2C. This makes dealing with I2C a lot more
     pleasant. While I2C theoretically supports 10-bit bytes, very few
     chips use this in practice.
  6. Simplify port code and interface to Erlang.
  7. Remove PWM interface. This was Raspberry Pi specific and not
     maintained.
  8. Make one multi-call binary. This reduces memory usage and
     simplifies the build system.
  9. Only support Linux. The previous code had an abstraction layer
     in anticipation of porting to RTOS's. That never happened, and the
     API had many Linux assumptions in it anyway.
 10. Support "atomic" write, read transactions on I2C. This is a very common
     usage scenario.
 11. Require client code to register names for ALE processes if it
     wants them. The previous code sometimes registered with gproc and
     sometimes locally. The user decides now.
 12. Remove most tests. They had limited coverage and complicated the GPIO
     code. The gpio_counter example still exists and has been updated.
 13. Use rebar3 to build.
 14. Fix gpio interrupt filter so that transients on the GPIO don't
     send an unexpected notification.

Signed-off-by: Frank Hunleth <fhunleth@troodon-software.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,66 +1,178 @@
...
-relatively straightforward too since the lowest layer hardware drivers are not
-baked into the ALE itself.
+Erlang/ALE provides high level abstractions for interfacing to hardware
+peripherals through I2C, SPI, and GPIOs on embedded platforms.
 
+# Getting started
 
-# Warning
+If you're natively compiling Erlang/ALE, everything should work like any other
+Erlang library. Normally, you would include elixir-ale as a dependency in your
```
Comment (knewter) [inline README.md]: references elixir-ale instead of Erlang/ALE here

File: README.md [hunk 2]
```diff
@@ -1,66 +1,178 @@
...
+
+    3> {ok, Gpio17} = gpio:start_link(17, input).
+    {ok, <0.97.0>}
+
+    4> gpio:read(Gpio17).
+    0
+
+    %% Push the button down
+
+    5> gpio:read(pid).
```
Comment (knewter) [inline README.md]: pid should be Gpio17

File: README.md [hunk 3]
```diff
@@ -1,66 +1,178 @@
...
+    {gpio_interrupt, 17, falling}
+    ok
+
+## SPI
+
+A SPI bus is a common multi-wire bus used to connect components on a circuit
+board. A clock line drives the timing of sending bits between components. Bits
+on the `MOSI` line go from the master (usually the processor running Linux) to
+the slave, and bits on the `MISO` line go the other direction. Bits transfer
+both directions simultanteously. However, much of the time, the protocol used
```
Comment (knewter) [inline README.md]: simultaneously

**Comment by fhunleth** [comment]
Fixed the README.md comments. I also noticed that the git clone line was pointing to my repo, so I changed that to the esl/erlang_ale repo.

**Pull request merged**